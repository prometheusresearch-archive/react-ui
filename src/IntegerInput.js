/**
 * @copyright 2016, Prometheus Research, LLC
 * @flow
 */

import * as React from 'react';
import invariant from 'invariant';
import Input from './Input';

import {tryParseInteger, extractValueFromEvent} from './FormUtils';

type Props = {
  value: string | number,
  onChange: (number | string) => *,
};

export default class IntegerInput extends React.Component<*, Props, *> {
  state: {
    value: string,
  };

  static stylesheet = {
    Input: Input,
  };

  static defaultProps = {
    value: '',
  };

  constructor(props: Props) {
    super(props);
    this.state = {value: props.value ? String(props.value) : ''};
  }

  render() {
    let {Input} = this.constructor.stylesheet;
    return <Input {...this.props} value={this.state.value} onChange={this.onChange} />;
  }

  componentWillReceiveProps(nextProps: Props) {
    let {value} = this.state;
    if (nextProps.value == null) {
      this.setState({value: ''});
    } else if (nextProps.value !== tryParseInteger(value)) {
      this.setState({value: String(nextProps.value)});
    }
  }

  onChange = (event: UIEvent) => {
    const value = extractValueFromEvent(event) || '';
    invariant(
      typeof value === 'string',
      'Expected a string value from the underlying input',
    );
    this.setState({value}, () => {
      if (value != null) {
        this.props.onChange(tryParseInteger(value));
      } else {
        this.props.onChange(value);
      }
    });
  };
}
