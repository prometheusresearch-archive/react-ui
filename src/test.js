/**
 * @copyright 2016+, Prometheus Research, LLC
 */

export function findByType(tree, type) {
  if (typeof tree.toJSON === 'function') {
    tree = tree.toJSON();
  }
  if (tree.type === type) {
    return tree;
  } else if (tree.children) {
    for (let i = 0; i < tree.children.length; i++) {
      let found = findByType(tree.children[i], type);
      if (found !== null) {
        return found;
      }
    }
  }
  throw new Error(`Couldn't find an element of type: ${type}`);
}

export function findAllByType(tree, type) {
  if (typeof tree.toJSON === 'function') {
    tree = tree.toJSON();
  }
  let found = [];
  if (tree.type === type) {
    found.push(tree);
  }
  if (tree.children) {
    for (let i = 0; i < tree.children.length; i++) {
      found = found.concat(findAllByType(tree.children[i], type));
    }
  }
  return found;
}
