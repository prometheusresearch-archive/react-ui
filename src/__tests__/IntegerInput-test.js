/**
 * @copyright 2016+, Prometheus Research, LLC
 */

import Renderer from 'react-test-renderer';
import React from 'react';

import {findByType} from '../test';
import IntegerInput from '../IntegerInput';

it('renders', function() {
  let onChange = jest.fn();
  let tree = Renderer.create(
    <IntegerInput onChange={onChange} />
  );
  expect(tree).toMatchSnapshot();

  let input = findByType(tree, 'input');

  input.props.onChange('12');
  expect(onChange).toBeCalledWith(12);

  input.props.onChange('12.');
  expect(onChange).toBeCalledWith('12.');

  input.props.onChange('12.2');
  expect(onChange).toBeCalledWith('12.2');

  input.props.onChange('xx');
  expect(onChange).toBeCalledWith('xx');

  input.props.onChange('');
  expect(onChange).toBeCalledWith('');
});
