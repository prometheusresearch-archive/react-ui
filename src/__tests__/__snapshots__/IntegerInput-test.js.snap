// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<input
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "backgroundColor": "#fff",
        "backgroundImage": "none",
        "border": Object {
          "color": "#ccc",
          "style": "solid",
          "width": 1,
        },
        "borderRadius": 2,
        "boxShadow": Object {
          "blur": 1,
          "color": "rgba(0, 0, 0, 0.075)",
          "x": 0,
          "y": 1,
        },
        "color": "#000",
        "display": "block",
        "focus": Object {
          "border": Object {
            "color": "#708698",
            "style": "solid",
            "width": 1,
          },
          "boxShadow": Array [
            Object {
              "blur": 1,
              "color": "rgba(0, 0, 0, 0.075)",
              "inset": true,
              "x": 0,
              "y": 1,
            },
            Object {
              "color": "rgba(0, 126, 229, 0.5)",
              "spread": 2,
              "x": 0,
              "y": 0,
            },
          ],
          "outline": "none",
        },
        "fontSize": "14px",
        "lineHeight": 1.42857143,
        "padding": Object {
          "horizontal": 12,
          "vertical": 6,
        },
        "transition": Array [
          Object {
            "duration": 0.15,
            "property": "border-coolor",
            "timingFunction": "ease-in-out",
          },
          Object {
            "duration": 0.15,
            "property": "box-shadow",
            "timingFunction": "ease-in-out",
          },
        ],
        "width": "100%",
      },
    }
  }
  disabled={undefined}
  onChange={[Function]}
  value=""
/>
`;
